Setup windows server as ad dc

piroor opened the pull request:
This adds some tasks to setup a Windows Server VM as an Active Directory Domain Controller. Those added tasks are skipped for non-server Windows clients.

yashirot commented:
Looks good but I will approve after trying whether that actually works.

yashirot commented:
One task failed in a test run.
Below is an excerpt from last messages:
```
TASK [Wait for the system to be ready after reboot for AD DC promotion] ************************************************
skipping: [4.241.169.203]

TASK [Setup chocolatey] ************************************************************************************************
fatal: [4.241.169.203]: FAILED! => {"changed": false, "module_stderr": "", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}

PLAY RECAP *************************************************************************************************************
4.241.169.203              : ok=27   changed=20   unreachable=0    failed=1    skipped=6    rescued=0    ignored=0


real    9m53.306s
user    0m11.560s
sys     0m4.514s
make: *** [../use-in-common.mk:18: apply] Error 2
```
Let's talk it over next business day.

yashirot commented:
It was good that AD DS was successfully there:
![image](https://github.com/user-attachments/assets/be4c8d54-4070-4cbf-844c-4823ce2c2c52)

yashirot reviewed (commented):
```
TASK [Setup chocolatey] ************************************************************************************************
fatal: [4.241.169.203]: FAILED! => {"changed": false, "module_stderr": "", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
```
Could you check over this failed task?

piroor commented:
The error around Chocolatey looks to be solved by changes in #59.

yashirot commented:
I will try it after merging PR #59.

yashirot commented:
A temporary test run failed with an error below:
```
TASK [Setup chocolatey] ************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: requests.exceptions.ConnectTimeout: HTTPSConnectionPool(host='74.176.170.216', port=5986): Max retries exceeded with url: /wsman (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x7f7ba3a729f0>, 'Connection to 74.176.170.216 timed out. (connect timeout=30)'))
fatal: [74.176.170.216]: FAILED! => {"msg": "Unexpected failure during module execution: HTTPSConnectionPool(host='74.176.170.216', port=5986): Max retries exceeded with url: /wsman (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x7f7ba3a729f0>, 'Connection to 74.176.170.216 timed out. (connect timeout=30)'))", "stdout": ""}

PLAY RECAP *************************************************************************************************************
74.176.170.216             : ok=29   changed=22   unreachable=0    failed=1    skipped=6    rescued=0    ignored=0
```

yashirot commented:
Running `make apply-playbook` ended up with another error: 
```
TASK [Set non-ASCII workgroup name] ************************************************************************************
fatal: [74.176.170.216]: FAILED! => {"changed": true, "msg": "failed to remove computer from domain: Failed to unjoin computer 'firefoxverify' from domain 'example.local' with the following error message: This machine is a domain controller and cannot be unjoined from a domain.", "reboot_required": false}

PLAY RECAP *************************************************************************************************************
74.176.170.216             : ok=3    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```

However this may not be a problem as I think ansible tasks are not idempotent differing from terraform recipes.

piroor commented:
> A temporary test run failed with an error below:
> 
> ```
> TASK [Setup chocolatey] ************************************************************************************************
> An exception occurred during task execution. To see the full traceback, use -vvv. The error was: requests.exceptions.ConnectTimeout: HTTPSConnectionPool(host='74.176.170.216', port=5986): Max retries exceeded with url: /wsman (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x7f7ba3a729f0>, 'Connection to 74.176.170.216 timed out. (connect timeout=30)'))
> fatal: [74.176.170.216]: FAILED! => {"msg": "Unexpected failure during module execution: HTTPSConnectionPool(host='74.176.170.216', port=5986): Max retries exceeded with url: /wsman (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x7f7ba3a729f0>, 'Connection to 74.176.170.216 timed out. (connect timeout=30)'))", "stdout": ""}
> 
> PLAY RECAP *************************************************************************************************************
> 74.176.170.216             : ok=29   changed=22   unreachable=0    failed=1    skipped=6    rescued=0    ignored=0
> ```

This looks happen due to unresponsive server for a while after reboot. We need to run it with some more delay.

piroor commented:
> Running `make apply-playbook` again ended up with another error:
> 
> ```
> TASK [Set non-ASCII workgroup name] ************************************************************************************
> fatal: [74.176.170.216]: FAILED! => {"changed": true, "msg": "failed to remove computer from domain: Failed to unjoin computer 'firefoxverify' from domain 'example.local' with the following error message: This machine is a domain controller and cannot be unjoined from a domain.", "reboot_required": false}
> 
> PLAY RECAP *************************************************************************************************************
> 74.176.170.216             : ok=3    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
> ```
> 
> However this may not be a problem as I think ansible tasks are not idempotent differing from terraform recipes.

Hmm. This error does not happen on my environment.

piroor commented:
After some retrying I've realized that Windows is restarted by Ansible itself and we need to wait until reconnected whether AD DC promotion is proceeded or not. The commit d8907bc applies "ansible.builtin.wait_for_connection" before further operations.

yashirot commented:
In a test run a subsequent task failed.
Below is an excerpt of PLAY RECAP

```
TASK [Promote to a domain controller if this is Windows Server] ********************************************************
fatal: [74.176.136.13]: FAILED! => {"changed": false, "cmd": "powershell.exe -command \"Install-ADDSForest -DomainName example.local -SafeModeAdministratorPassword (ConvertTo-SecureString 'jaGd,CC|0CR]TeeA~u-Q' -AsPlainText -Force) -Force\"", "delta": "0:00:02.531170", "end": "2025-05-01 05:06:05.210264", "msg": "non-zero return code", "rc": 1, "start": "2025-05-01 05:06:02.679094", "stderr": "Install-ADDSForest : Verification of prerequisites for Domain Controller promotion failed. The specified argument 'NewD\r\nomain' was not recognized.\r\nAt line:1 char:1\r\n+ Install-ADDSForest -DomainName example.local -SafeModeAdministratorPa ...\r\n+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\r\n    + CategoryInfo          : NotSpecified: (:) [Install-ADDSForest], TestFailedException\r\n    + FullyQualifiedErrorId : Test.VerifyDcPromoCore.DCPromo.General.77,Microsoft.DirectoryServices.Deployment.PowerSh \r\n   ell.Commands.InstallADDSForestCommand\r\n", "stderr_lines": ["Install-ADDSForest : Verification of prerequisites for Domain Controller promotion failed. The specified argument 'NewD", "omain' was not recognized.", "At line:1 char:1", "+ Install-ADDSForest -DomainName example.local -SafeModeAdministratorPa ...", "+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~", "    + CategoryInfo          : NotSpecified: (:) [Install-ADDSForest], TestFailedException", "    + FullyQualifiedErrorId : Test.VerifyDcPromoCore.DCPromo.General.77,Microsoft.DirectoryServices.Deployment.PowerSh ", "   ell.Commands.InstallADDSForestCommand"], "stdout": "\r\nMessage                                                                                                                \r\n-------                                                                                                                \r\nVerification of prerequisites for Domain Controller promotion failed. The specified argument 'NewDomain' was not rec...\r\n\r\n\r\n", "stdout_lines": ["", "Message                                                                                                                ", "-------                                                                                                                ", "Verification of prerequisites for Domain Controller promotion failed. The specified argument 'NewDomain' was not rec...", "", ""]}

PLAY RECAP *************************************************************************************************************
74.176.136.13              : ok=32   changed=24   unreachable=0    failed=1    skipped=5    rescued=0    ignored=0


real    16m12.307s
user    0m18.669s
sys     0m6.422s
make: *** [../use-in-common.mk:18: apply] Error 2
```

yashirot commented:
>  The specified argument 'NewD\r\nomain' was not recognized.

Seemingly an unintentional \r\n is inserted.

piroor commented:
I've rebased the branch and tried to run. The error commented at https://github.com/clear-code/firefox-support-common/pull/56#issuecomment-2844126649 does not happen on my case. Hmm...

```
TASK [Promote to a domain controller if this is Windows Server] ***********************************************************************************
changed: [4.189.232.51]

TASK [Restart after promotion if this is Windows Server] ******************************************************************************************
[WARNING]: A scheduled reboot was pre-empted by Ansible.
changed: [4.189.232.51]

TASK [Wait for the system to be ready after AD DC promotion] **************************************************************************************
ok: [4.189.232.51]

TASK [Download Microsoft .NET Framework 4.8 Offline Installer] ************************************************************************************
changed: [4.189.232.51]
```

yashirot reviewed (commented):
Tried again and a `make all` went through successfully.
I can see the AD DS deployed on the Dashboard of Server Manager.

![image](https://github.com/user-attachments/assets/134cc5a3-c68b-4a06-9030-427bf834fc97)
